verify/cert.go

```diff
@@ -49,40 +51,60 @@ func NewCertVerifier(cfg *Config, l log.Logger) (*CertVerifier, error) {
 	}, nil
 }
 
-// verifies on-chain batch ID for equivalence to certificate batch header fields
-func (cv *CertVerifier) VerifyBatch(
-	header *binding.IEigenDAServiceManagerBatchHeader, id uint32, recordHash [32]byte, confirmationNumber uint32,
+// verifyBatchConfirmedOnchain verifies that batchMetadata (typically part of a received cert)
+// matches the batch metadata hash stored on-chain
+func (cv *CertVerifier) verifyBatchConfirmedOnchain(
+	ctx context.Context, batchID uint32, batchMetadata *disperser.BatchMetadata,
 ) error {
-	blockNumber, err := cv.getConfDeepBlockNumber()
+	// 1. verify batch is actually onchain at the batchMetadata's state confirmedBlockNumber
+	// This assert is technically not necessary, but it's a good sanity check.
+	// It could technically happen that a confirmed batch's block gets reorged out,
+	// yet the tx is included in an earlier or later block, making this check fail...
+	confirmationBlockNumber := batchMetadata.GetConfirmationBlockNumber()
+	confirmationBlockNumberBigInt := big.NewInt(0).SetInt64(int64(confirmationBlockNumber))
+	_, err := cv.retrieveBatchMetadataHash(ctx, batchID, confirmationBlockNumberBigInt)
 	if err != nil {
-		return fmt.Errorf("failed to get context block: %w", err)
+		return fmt.Errorf("batch not found onchain at supposedly confirmed block %d: %w", confirmationBlockNumber, err)
 	}
 
-	// 1. ensure that a batch hash can be looked up for a batch ID for a given block number
-	expectedHash, err := cv.manager.BatchIdToBatchMetadataHash(&bind.CallOpts{BlockNumber: blockNumber}, id)
+	// 2. verify that the confirmation status has been reached
+	// We retry 5 times waiting 12 seconds (block time) between each retry,
+	// in case our eth node is behind that of the eigenda_client's node that deemed the batch confirmed
```

Comment:
why would we ever use two nodes? or is the idea to have some resiliency to transient failures where the provider could be unreliable due to poor configuration?

Reply:
Both eigenda-client and proxy dial the endpoint individually, establishing independent connections. If the nodes are behind a load-balancer, then both connections could actually be connected to different eth-nodes, its very frequent that rpc provider nodes are not in sync (bit me very hard once) for some dumb reason (web3 standards...).

Even if we injected the dialed connection from the proxy into the eigenda-client, the underlying tcp connection could drop in between the eigenda-client returning and proxy making its call, which would reconnect to another node, with some effect.

Reply:
makes sense and was thinking that too - got burnt in the past using a node cluster with an ALB ingress where subsequent RPC queries could fail due to nodes being out of sync. Ik OP clusters typically use [proxyd](https://github.com/ethereum-optimism/infra/tree/main/proxyd) to mitigate this from happening. from what I remember the connection would be held with the ALB directly with different nodes being queried per each new request on the client --> ALB connection. Feel free to resolve this comment!

Reply:
Actually on second thought - isn't this a bit naive when we have access to the `L1ReferenceBlock` number at which the batch should've confirmed? Also we're adding a 12 second cool-off which is aligned with the block production interval on Ethereum but not syncing

Reply:
proxyd doesn't seem to solve this problem, they say: "we consider healthy is... latest block lag ≤ configurable threshold". Way to solve this is to use sticky sessions.

Not sure I understand your last comment. Is https://github.com/Layr-Labs/eigenda-proxy/pull/192#discussion_r1818085482 related to what you are talking about?

Reply:
Talked offline. Let's update to retrying every 2 seconds for the same amount of time (60 seconds).

Reply:
[0c54503](https://github.com/Layr-Labs/eigenda-proxy/pull/192/commits/0c545036bcf8f27cb9dac9882aa8649fd508a659)